### ISSUE TITLE
New LCD proposal

### ISSUE THREAD

**Issue by sp9bsl** [df8oe/mchf-github#1089]
Hello everyone, my name is Slawek, this is my first post in this topic. 
Encouraged by Andreas, i would like to present some work with raspberry pi LCD connected to UI board.
This LCD module has resolution 480x320, size of 3.5" and SPI interface.
You can see it working in my video published few weeks ago: 
[video](https://www.youtube.com/watch?v=HzEP_i8pxCs&t=41s)
I made simple adapter board with few connectors and 5V voltage regulator, simply inserted in socket for 2.8" LCD. The only modification of UI board is connected 8V for 5V voltage regulator on unused LCD pin. I decided to add a voltage regulator because of thermal hazard of U3 (LM2941), this LCD (its backlight) needs about 80mA more current. This is all regarding hardware. Ofcourse i can always step back to old 2.8 LCD by simply replacing it without soldering. This LCD is also few milimeters bigger in both directions and won't fit in tight case prepared for current LCD.

Now about firmware. 
As you can see on video there is simultaneous spectrum and waterfall display. This is the main advantage over normal solution with 2.8" LCD. I managed overall outfit of lcd by watching video recorded by HAM  from Russia:
[video](https://www.youtube.com/watch?v=PrIvSIxiitw&t=26s)
As you may think, there is a lot of differences with driver code. This LCD uses ILI9486 controller and it has little similarity to ILI932x, so almost all basic functions regarding communications required changes.
Also because of twice as width compared to old LCD i also changed buffers size and depth of FFT.
Simultaneous spectrum and waterfall display requires more memory for buffers that are swapped between waterfall or spectrum in current firmware.

Display of spectrum and waterfal at one time required rewrite DMA based LCD functions to speed it up.
Instead of drawing every bin of FFT pixel by pixel i changed it to two pre drawn buffers. The first buffer contains background with grid, second is full filled with color of spectrum display also with grid. Drawing function simply changes buffers instead of pushing every pixel to buffer and flushing it when is full. This step gives speed up for about twice. I also added highlight for filter bandwidth working in the same way.

There is one drawback, i did not test freedv communication. Display improvements i made required more RAM for double buffers for FFT results needed by scope and waterfall. Now i have 102292 bss and 3736 data. There is some room for heap/stack but i cannot guarantee that freedv will now work, . I didn't mention that i still use STM32F405 with 192kb of RAM. Although i didn't find any RAM usage statistics for freedv.

I know that it will require a lot of work to integrate new lcd with current project, although i think that this change will bring big enhancement in overall look on this fantastic project. There is a lot to speak and think about it, i'm open for any suggestion where to start. At the beginning there should be a little cleanup with current definitions of placement items on LCD and touch. Plenty of them are coded inline (touch panel mainly). If we decide to integrate this lcd to current project, i think there should be conditional code compilation for new lcd version because of different item positions and size.

73, Slawek SP9BSL

**Comment by s53dz** [comment]
I wonder if in the case of introducing another/bigger LCD it would not be beneficial to add also a dedicated MCU for operating LCD. Just guessing. Probably in this case an STM32F429VG with 1MB flash and 256kB RAM could do its mcHF SDR job just fine.

73 Bojan

**Comment by df8oe** [comment]
Hi Slawek,

your suggestion and improvement is welcome. Your idea to implement it as "conditionally building" is very good, too. Freedv is eating much memory - there is not much free RAM at the time if you use STMs with 192KB of RAM. Maybe there can be done some enhancements - but I am not sure how much RAM this can free. Additinally, UHSDR works with STMs which do have 256KB of RAM (already in mcHF) - and on OVI40 there is F7 with 512KB of RAM. So the step
- adding code so that it is built conditionally
frees up to do possible dynamic handlings later.

Regarding supply current: The 3.2" LCD which I use on adaptor PCB does consumpt ~60mA more than 2.8" - and is working using old regulators. Of course concept of 8V regulator is not a lucky one (and will be completely different in OVI40) - but it will still work if you do have good cooling.

Regarding of ui placement:
320x240 LCD of course will have another layout as 480x320. So there must be two different uis. We must talk about functions which can be used by both ui types so that code duplication is minimized. We must talk about the way we will start this. You have done your tests using nearly new firmware versions so you can make proposals what must / should be changed first.

vy 73
Andreas

**Comment by yo2ldk** [comment]
I personally, prefer the variant with both spectrum and waterfall, freedv is completely useless for me, and certainly for many others.
  Also, based on OVI40, I dream for a good 5" LCD (as it is on STM32F7 development boards) , at least on OVI40-II  :)

   73 de alex - yo2ldk

**Comment by df8oe** [comment]
OVI40 ui is very flexible and can (and will) support bigger LCDs. But for portable use it still can handle 3.2" LCD and this is the one which will be the "default one". Bigger LCDs are simply coupled to SPI and can be added later.

73 de
Andreas, DF8OE

**Comment by sp9bsl** [comment]
Dedicated microcontroller for LCD is good idea but this is completelly new design and total change in the code. Better way for this project is to change microcontroller to F7 and make benefit of its SIMD instructions. Unfortunatelly i cannot find info/schematics of OVI40 trx - can you point me where should i look to find it? My hardware is build on mcHF boards so simple changing F4 to F7 microcontroller is not possible (pin mismatch). So now i can only work with small improvement using adapter pcb converting 2.8" to 3.5" lcd on the ui board. New socket for LCD is highly welcome but this is Chris part of work due to licensing (i think). Besides 3.5" lcd is the best price to functionality ratio. Ofcourse we can use dedicated gfx controller from 429 of F7 with SDRAM and it should help a lot with core load, but again it will require board redesign. There is also possibility to use assembly language (i use it often in M3/F4 designs). 
Now about freedv: i never hear nor have QSO with freedv so for me it is also useless, but some people may think different and i have respect for that.
Andreas, i think that most code will be the same for both types resolution of LCD, the most differences will be at positioning of items and buffer lengths (for example FFT length). Now i should send schematic and pcb for cloning my simple board, where can i do it?

**Comment by db4ple** [comment]
Hi,
let us go step by step: What we have to do is the following:

1. We want to implement a higher resolution spectrum (i.e. larger FFT using more RAM).
1. We want to support a new graphics controller, in addition to the existing ones. In an ideal scenario the "highlevel" interface (the functions called from other files to do the actual user interface are identical) so that the only thing which differs  is the implementation of the hardware access (the driver implementation).
1. We want to support an alternative resolution in addition to the existing one. As you pointed out, this requires rearrangement of ui elements or even resizing. Taking advantage of larger space would permit placement of more elements on the screen (scope + waterfall).

First item is relatively simple unless we are running out of RAM. Leaving out FreeDV for the low RAM scenarion would save 64k and that should be plenty for twice as large fft (which increase RAM use by about 12k according to your numbers and also my calculation looking at the data structures). Enlarging the waterfall from 256 to 480 would also increase the RAM usage unless we can follow a more clever approach if the graphics controller does screen buffer shifting. Then we can save a lot of RAM and performance (this is in fact currently the most time consuming ui operation when completely redraw the waterfall for every scrolling step).

Second item: well you did that, we should look at the code to see how feasible a uniform interface is, but I think this is also not too difficult. 

Last item is the most complex one since the graphics layout is currently deeply embedded in the code and we need to consolidate that in a better way to have multiple layouts for different resolutions without too much effort.

First step: published your code so that our discussion can be based on facts rather than assumptions.

73
Danilo

**Comment by sp9bsl** [comment]
Hi Danilo, i will publish code at evening.
73 Slawek

**Comment by df8oe** [comment]
I will open a new branch and call it lcd-experiental

vy 73
Andreas

**Comment by df8oe** [comment]
Finished - new branch is available.
We can comment out freedv for lcd implementation of course. What happens later: we will see. There are more options than one and minimum two different builds (one for OVI40 and one for mcHF).
Here you can find schematics of ui-OVI40 board: https://www.amateurfunk-sulingen.de/data/uploads/ovi40-sdr/uhsdr_ovi40_ui_v17_sheet.pdf

I will create webpages regarding OVI40 at the weekend.

73
Andreas

**Comment by db4ple** [comment]
Hi,
it is currently not possible to comment out FreeDV totally (this was broken) but I will fix that. Meanwhile it is sufficient to disable FreeDV in menu and to comment out the FreeDV setup call in the ushdr_main.c.

**Comment by sp9bsl** [comment]
I will send changed files to Danilo and Andreas to show you what i did. Github style of coding is new to me and i must learn some, so this will be faster for now. I will start to add code to new branch in next few days after some reading and learning. Thanx for schematics of OVI40 board, now i see big enhancements by using two codecs and F7 still maintaining compatibility with RF board which also needs rework :)
73 Slawek

**Comment by df8oe** [comment]
@Danilo:
Yes - I remember... We will add theese changes to both branches as long as they are equal to have a "good start".

@Slawek:
For the first time this is ok of course. But getting familar with git (and GitHub) is appreciated because it can save many, many hours of (merging) work and making community working comfortable...

73
Andreas

**Comment by df8oe** [comment]
Do you work on Windows or Linux? Of course will help you getting familar with GitHub.

**Comment by sp9bsl** [comment]
Andreas, you're right i will go this way. I work with Windows and Eclipse.
73 Slawek

**Comment by df8oe** [comment]
Slawek, I am working with Linux mostly on command line.... So Danilo is the one who knows Windows specific things. But I have read on Yahoo forum that there is a GUI which supports all OS, supports most technics of git and is free - I think you should have a look at it: gitkraken

73
Andreas

**Comment by db4ple** [comment]
Hi Slawek,
before we integrate your changes in whatever way, just try to simply create a new project in GitHub with your content, it does not have to be connected or forked from the UHSDR project.

Then we can start integrating these changes into the lcd-experimental branch step by step. We will help you to get used to Git in general and how we use it in combination with GitHub.

73
Danilo

**Comment by sp9bsl** [comment]
I think i did it. [Here](https://github.com/sp9bsl/mcHF) you can find code as is now in my radio shown in first post here. Now i see that some of the comments are in my native language (polish) sorry for that.

The list of files i made changes:
basesw\mcHF\Src\spi.c
drivers\audio\audio_driver.c
drivers\audio\audio_driver.h
drivers\audio\freedv_mchf.c
drivers\ui\ui_driver.c
drivers\ui\ui_driver.h
drivers\ui\lcd\ui_lcd_hy28.c
drivers\ui\lcd\ui_lcd_hy28.h
drivers\ui\lcd\spectrum.c
drivers\ui\lcd\spectrum.h
drivers\ui\menu\ui_menu.h
drivers\ui\menu\ui_menu_internal.c
hardware\mchf_board.c
hardware\mchf_board.h
hardware\mchf_board_config.h

73 Slawek

**Comment by db4ple** [comment]
Great. Will have a closer look at it today. I will see how we can fast forward your changes to a certain degree.

I would propose to start with integrating the low level graphics driver trying to get the unchanged layout to work on the larger screen (will look strange I guess), this looks like a good candidate. I might even have a ILI9486 based display 3.5" RPI display like yours, not sure though, need to check.

73
Danilo

**Comment by sp9bsl** [comment]
Yes, this is the right way. I did it like you said. First change the display and low level routines on current layout. There is also need for SPI edge/phase change compared to old LCD, i spent two days on it with white screen before notice what is wrong. Look at my code regarding spi settings also. My LCD is Waveshare clone. Danilo, will you try this or this is my task?

73 Slawek

**Comment by s53dz** [comment]
Hi Slawek,

I welcome your proposal very much since I think that 2.8" LCD is quite small for my fingers, hi. 

I am observing some inter-modulation distortion resulting from the rate of 2.8" HY28B LCD refresh cycle. With only UI board connected it is presenting in the range of 14.014 to 14.017 MHz, for example. Although with a magnitude of only -120 dBm but still.

I suppose a 3.5" LCD project is only out of a baby phase now. But I wonder if you have measured or observed the inter-modulation or distortion of this kind from your 3.5" LCD? I suppose with better resolution also the refresh rate would be higher there.

**EDIT:** I know this is a HW question, however I am curious what to expect.

TNX 73 Bojan

**Comment by sp9bsl** [comment]
Hi Bojan,
this LCD is only a little bigger, but pixel size is smaller because of resolution (twice more pixels than in 2.8"). Items are a bit smaller than presented on 2.8". I just use a stylus from old HTC phone to point them by touch panel, also when i buy this lcd i get plastic stylus in the package.
Regarding interferences: i see something in freq range you asked, it looks like some kind of alias on -115dBm, but i don't see any other signal near except peak at 14.000 with amplitude of -100dBm. I didn't care about that coz i have background from my antenna(W3DZZ) at level of S7. 

73 Slawek

**Comment by sp9bsl** [comment]
Should i start integrating low level routines or wait? I don't want to go in anyone's scope of authority.

73 Slawek

**Comment by df8oe** [comment]
Hi Slawek,

we do not have any authority - we only have to coordinate all works so that it gets best performance and no loss because of two persons are working at the same time on similar parts of the code...

So let us discuss how we will do it.:
I have created new branch lcd-experimental
You should clone and use this branch for implementation.
@Danilo: do you have begun with work / how will we divide it? I will not distrub integration.

But I have got 3.5" LCD and will fit it to my OVI40-UI (SPI lines are already on the bus - just define one more for CS_LCD and everything can start up ;)

73
Andreas

**Comment by db4ple** [comment]
Hello all,
if this is okay with all of you, I would like to propose how to make progress. This would require me to prepare a few things in the code and explain you, Slawek, how we organize this. I can unfortunately start this no earlier than tonight (i.e. not now). Appreciate you waiting. Basic idea is to use the feature branch lcd-experimental as Andreas suggested.

73
Danilo

**Comment by sp9bsl** [comment]
OK, no problem i will wait.

73 Slawek

**Comment by db4ple** [comment]
## First step

I created in my (db4ple) GitHub Repo a new branch (lcd-ili9486) which "starts" at release 2.2.0 and has your changes (most of them) as next commit. This gives us a nice overview of what you did  change in your effort:

https://github.com/db4ple/mchf-github/commit/daba61987dd11b7188d7a162f2295c06cfbbc03b

## The setup
Now fastforward to the current activities:

You setup your git/github environment so that you can publish your changes to the DF8OE repo.
Please use these instructions: https://github.com/db4ple/mchf-github/blob/active-devel/CONTRIBUTING.md
I made small changes to the text there after looking at them (will be soon merged into the "official" repository). In general the CONTRIBUTING.md from DF8OE should be looked at...

Instead of using 'git checkout active-devel' you should use 'git checkout lcd-experimental'.
In git rebase, do not replace active-devel with lcd-experimental
When you do the pull request, use lcd-experimental as target branch.

You can try that workflow with a small commit to see how this works before starting the real work.

## The real stuff
Looking at the changes, I would propose to use 3 flags

For step 1:

USE_GFX_ILI9486 to identify the places, where you make changes do to the new driver which are independent from the resolution (these are the changes we do in first step, only these please).

For step 2/3:
USE_DISP_480_320 to identify places where the screen resolution related changes are addressed.

USE_FFT_LARGE to identify places where the FFT size needs to be changed. Maybe we just use the FFT SIZE but I don't know yet, we'll see.

Please keep in mind: This is a proposal from my side, if you have different ideas, just tell.


## The goal
We integrate your change so that we can (by means of conditional compilation) build the "normal" mcHF firmware and the other one with ili9486 480x320 display (if the defines are set accordingly).

**Comment by sp9bsl** [comment]
Thanks Danilo, it is OK i think. Thanks for commiting this to v2.2.0, now all can clearly see what has been done.
One question: by saying "3 flags" you mean #ifdef in those places right?
I tried your instructions link, but it is not working.

73 Slawek

**Comment by db4ple** [comment]
I fixed the link. 
a) you can define the new flags in uhsdr_board.h (you will see, there other USE_...  already there).
b) and yes, then we use #ifdef USE_....

**Comment by DG9BFC** [comment]
hmmm.. if we swap to the mcu with abit more ram we maybe could use the bigger display AND let freedv inside the code??? ... is the adaptor pcb for that bigger display available??? i would like to have that bigger screen resolution (and spectrum and waterfall on screen)

**Comment by db4ple** [comment]
Hi Slawek,
just to be clear: Go ahead and add the driver to the original code in the lcd-experimental branch  once your github setup is working. If you have trouble with git/github let us know.
73 Danilo

**Comment by db4ple** [comment]
@DG9BFC: I don't think there is a general problem to run all of this together once the RAM requirements are met. So for larger RAM devices FreeDV should work as before even with the new display. Of course, a larger FFT is also consuming more computational power and more data needs to be sent to the screen so the max update screen rate may not be as high as before.

**Comment by sp9bsl** [comment]
Hi Danilo,
sorry for delay, i have some weekend activities, and also urgent task for my employer. I will start today at evening.
73 Slawek

**Comment by DG9BFC** [comment]
at slawek ... do you have any more of the adaptor boards (maybe with display so plug and pray?!?)

best would be if the fw would detect the needed settings at boot and so you can start with small or big display (shut off mchf ... plug in bigger display and restart) ... but i am guessing that would not work (or would boost up firmware size tooo much) .... i really like that new view (as seen in the video) and the better screen resolution is superb (especially if you ad more digimodes later that better resolution is fine i guess)

i would like to have such a display and try out the "new lcd experimental fw"

really looks great

maybe for original mchf fw some of your clever display driving tricks could be used to speed up the display (and also change layout to have wf and scope)?!?

greetz
sigi

**Comment by db4ple** [comment]
@sp9bsl:  No worries, take your time.  Just wanted to make sure, you are not stuck. 
@DG9BFC: While the dynamic selection of the screen layout seems to be nice, this is a much larger piece of work due to the existing architecture. And if this is done wrong, it will become a software maintenance nightmare.  So I suggest to static to the static compilation. 
The list of reasons for not doing it include (don't get me wrong, all of this is doable): 
  * Dynamic Sizing / Memory allocation for FFT without using extra memory on low memory machines. Low memory machines may not be able to run the large display although they detect it, if they want to use FreeDV...
  *  The layout (x and y coordinates of UI elements) are current "compiled in", making the dynamically switchable needs some attention to the details too.
* Effort spent on the to previous items makes sense if users would really switch displays often. Which they don't typically  I would assume.

**Comment by sp9bsl** [comment]
DG9BFC: sorry i haven't any pcb nor lcds now. I did it in home - it's prototype.
Danilo explained  everything about lcd autodetection, i must admit that this LCD has one way SPI communication, there is no possibility to detect it, only touch panel chip has back communication.

73 Slawek

**Comment by K2AAE** [comment]
I was looking at another issue related to this issue: 
1. Move the Touch Screen Code to separate file, currently there is no means for Capacitance I²C touch screens.

2. Dynamic build system with users checking boxes to get the firmware for their hardware, did that for ESP8266.

3. The BIG Picture plan on LCD support moving forward. While the F7 can support external memory the H7 goes further with more graphics hardware support both of which isn't supported by F4xx. 

4. Virtual screen, data sent to external display.

As it stands I found some 3.5" displays with either Resistive $2 or Capacitance Touch Screens $5, ILI9488. 320X480 no daughterboard, also some 3.5 320X240 SSD2119 with Touch screens, no daughterboard and 320X240 3.2, ILI9341, Resistive $2 or Capacitance Touch Screens $5, no daughterboard. To make a daughterboard with the pins to mate the UI for these is $3. They all have one thing in common: no display API, everyone of these controllers has to be beat into place with time and after looking at the existing code I figured we'd need to talk and you already are.

At any rate I'm still waiting on the first 2.8" display I ordered a month ago from hotmcu.com, it took 8 days to ship and he must of put it on a very slow boat. The bare boards I ordered from Chris arrived but the stuffed boards are still traveling so I basically have no hardware to test with yet, I had to take a break from rebuilding the parts list. 

I did start on a shield for the Nucleo H746 that will mate with the RF board, have all the switches and controls along with Si5351 and WM8994, something to do while I wait on the slow boats.
Kevin K2AAE

**Comment by db4ple** [comment]
Hi Kevin
1. Yes, once it gets to messy we should move it out. I think main motivation here was that both display and touch may share the SPI if both are connected using the SPI bus and thus have kind of a tight link. Due to maximum performance based on optimization and inlining the code was kept in a single file.
By using gcc's LTO  (which we do since quite a while) this may no longer be true and we can separate code easily without impacting performance. 

2.  Make an independent issue for that and suggest how we could do this, I am very much in favor of that idea.

3. and 4. I am not exactly sure what you mean. 

Honestly, I don't think we should try to support every controller out there. However, moving to a different resolution is a somewhat logical and necessary step.  Yet we have to keep in mind the intention is to keep the current mcHF HW in the loop and not to abandon them to quickly.  
All of that puts some constraints on us. The current software architecture has not been made with hardware flexibility in mind, so it will be some kind of longer process of evolution to get there. 
Having 2. in place (or some equivalent of this) gives us some freedom to work on the evolution I would say, since it will permit us to provide more configurations without need to build all of them ahead of time and we can also track which configurations are being used and which not.

**Comment by df8oe** [comment]
I have already thought about a webinterface and the possibility to build special configurations on my server. I would buffer each build and if same is selected again it would caught from cache. So we can possibly provide more flexibility (of course I cannot archivate huge amounts of different builds on my server due to storage limitations). As Danilo already wrote: our intention is _not_ to spread into every direction which is possible in theory (and reality) - we want to provide a support of a selected hardware. If some else wants to support special hardware configurations he can create his own branch (or, if interest is big enough, we will think about supporting it inside UHSDR main repository). But it is not planned and will not be done to support "special" hardware configurations which will not present in "nature" in a minimum amount.

vy 73
Andreas

**Comment by sp9bsl** [comment]
Andreas: this lcd (3.5" from raspberry i use) has 2 shift registers which convert SPI to parallel interface. The back spi node is only for touch panel IC.

73 Slawek

**Comment by df8oe** [comment]
Yes - I do have the same one. It has four 16pin COIC16 on back. Maybe the LCD is available in parallel wiring, too. But all this is subject of the future. Interesting is that the form factor of the LCD by itself has exactly the same form factor of the 3.2" we use at OVI40 if you include the PCB where it is mounted... So it is mechanically "drop in compatible" you only have to rewire it...

**Comment by db4ple** [comment]
Hi, my 3.5 RPi also uses a shift register. This is very interesting since the ILI controller can do SPI just fine. 
BTW, I just broke mine in an attempt to figure out which controller chip is on, I killed the flexcable beyond repair. I ordered a new one. 

For using it with the mcHF and a newly developed adapter: I think this display could be right since it has the flexcable connected using a proper connector. Here an adapterboard could be more easily be developed for. 

http://www.ebay.de/itm/Waveshare-3-5-inch-320x480-Touchscreen-TFT-LCD-Display-for-Raspberry-Pi-WS40003-/281842700934?hash=item419f225286:g:I7wAAOSwviBZxOwg
 
73
Danilo

**Comment by DG9BFC** [comment]
hello danilo ... would it be possible to add those raspi headers direct to the i40 ui board?

just in the layout with solder spots (so rewiring would be with some soldering work to be done)

or best with connecting the header in the layout in parallel with existing one?!? (a massive rework on the layout to be done ... so not that good)

my dream would be that you can plug in different lcd screen (take what is available in evilbay) and fire up the right fw (taken from andreas server ...) ... and have a working ui

i do not know if we could add some different screen headers in the layout (and rewiring has to be done from the user but i think it is not possible to have all different lcd screens (with its different header locations) on the ui board ... so ... to be flexible (also for later development) we have to have a display header ... and one (or a few) adaptor board(s) .... guessing that is the best way to have:

different screen sizes and resolutions available

and if that screen where i40 sdr is built and planned with is not available anymore in a few years (we have a good example with the 2.8 inch display) ... just plug in a different one

**Comment by df8oe** [comment]
Hi Sigi,

design of ui PCB is finished and will not be touched now. You can add new LCD (the original Raspi one) by adding 8 wires. It has exact same form factor as 3.2" LCD. Maybe late there will be a "special LCD" which carries the original display but does have a self designed PCB which adapts parallel mode using exactly the same connection as 3.2" has. But all this is future stuff. First we want to get a working device using the parts where PCB is designed for - rest is done in "next steps" ;)

**Comment by DG9BFC** [comment]
just thought about future (and what to do when a display becomes unavailable?) ... I DID NOT WANT TO CHANGE LAYOUT NOW FOR SURE ... if that was understood that way ... sorry
if adding the new display takes only a few wirings to be done ... ok on that ... thats fine (and not more work as touch screen mod on mchf)

**Comment by df8oe** [comment]
The choosen 34pin header for LCD is as flexible as it can be. It contains all voltages and signals for parallel and SPI driven LCDs.

**Comment by sp9bsl** [comment]
Danilo, i have problem with push. What exactly command should i use? When i do `git push origin lcd-experimental` i get Everything up-to-date, when i try `git push main lcd-experimental` i get permission denied.

73 Slawek

**Comment by K2AAE** [comment]
Nice reading on this problem: https://goo.gl/k5iwdf 
I think it would be a good idea to set some expectations for going forward:
1. Current MCU can support? 320X240/480X320 with GRAM on the LCD
2. Which controllers to support: ILI932x + ??? ILI948x?? Build time selection? 
3. Develop a shield for this type of display: https://goo.gl/khwoLj -  SPI for LCD+SD Card+Flash Font memory+Touch+I²C for Touch on one flex cable with ZIP style connector, or this: https://goo.gl/EiVUbU at 5AMP per pin it could handle the display, I'm waiting on Molex, they left "number of insertions" out of the specs. Of course use a 74LVC139D to free up some pins. Add connectors that can be soldered for the OVI40 and mcHF but I haven't seen the OVI40 layout so I don't know if this is possible. 
With this those that had mcHF could upgrade to Flash Fonts and larger display "cheaply", could also make a 2.8" version.  For this application (mcHF footprint) and the small displays SPI is very usable from what I can tell. 

Since the current design can only 'fit' a 3.5" display (I used the front panel in CAD to find out what would fit, 4.3" height wise with moving all the controls to the left) and there is 320X240 3.5" display as well as 320X480 3.5" displays that looks to be the limit for the mcHF style. To move to the next generation of MCU with built in LCD support is another project since there isn't the room on the mcHF for a bigger display and the MCU is a bottleneck.

NPX has a M4 part that can handle 1024X768 $20 and new layout. About the time you get done you discover the LCD vendors are offering HD, 2K and 4K MIPI-DSI glass. 

I feel the small and simple is the path to take on this, I don't want a Flex ten zillion like radio.. Using IceStorm opens some interesting projects.

**Comment by db4ple** [comment]
@Slawek: All of this is normal:
- 'main' is Andreas GitHub repository, you can read from it but not write to it. 
- 'origin' is your GitHub repository, you can read from and write to it. 
- Since you already pushed your local repository's to your GitHub repository (aka 'origin') already, a second push will tell you this.

What you have to do is:
- go to your GitHub repository ( https://github.com/sp9bsl/UHSDR )
- press New Pull Request
- Select the lcd-experiemental branch in the df8oe repository (this is the left side of the picture below) instead of the default "active-devel". Then you create the pull request.

![2017-10-02_08-38-16](https://user-images.githubusercontent.com/17237769/31085667-187697ce-a798-11e7-9c46-b5bbea30ed45.jpg)

This should be it...,

**Comment by db4ple** [comment]
@Airheadbit : I think we are aligned here. While there is some additional interest in getting a 5 inch 800x480 display connected, this is not an update option for current users with a "normal" mcHF UI like experience as this cleary requires a new UI layout or a separated display.

**Comment by sp9bsl** [comment]
Some mechanical drawings (3D view) how the 3.5" LCD fits in my case.

73 Slawek
[my mcHF inside.pdf](https://github.com/df8oe/UHSDR/files/1349494/my.mcHF.inside.pdf)
[my mcHF outside.pdf](https://github.com/df8oe/UHSDR/files/1349495/my.mcHF.outside.pdf)

**Comment by K2AAE** [comment]
@sp9bsl checkout https://goo.gl/xDQMvQ He has done a nice job on the ILI932x/934x/948x controllers
It seems the ILI9481 is faster than the ILI9486 and the ILI9488 is the latest. Code to look at in MCUFRIEND_kbv.cpp

**Comment by K2AAE** [comment]
@sp9bsl  a couple of empty PDF files! 
![screen](https://user-images.githubusercontent.com/7192199/31087879-f8752f26-a75b-11e7-9883-f70d4a685db1.jpg)

**Comment by sp9bsl** [comment]
Hi Kevin,
thanks, now i will add some driver code. It has no influence on speed. The real speed improvement will arrive when we start to change ui code. The real power of this lcd is limited by hardware (maximum spi speed on slow APB1 bus). Regarding files : just download pdf's and open them outside of browser (like in Adobe Acrobat), fileas are 3D i doubt your browser can open it.

73 Slawek

**Comment by K2AAE** [comment]
@db4ple The 800X480 is the maximum size display the i.MX7D can handle, dual ARM 7 MPU @1200Mhz with a M4 at 200Mhz for real-time tasks. The i.MX8 is suppose to have GPU's at about the same price point $20.

**Comment by db4ple** [comment]
Hi guys,
there is a big speed improvement to be had with the ILI9486 controller which is not possible with the ILI932x controller family and that is easy vertical scrolling. This will reduce memory footprint of the waterfall to just one line (the current one). The remaining waterfall will simple be scrolled one line with a single controller command. That means we can have both at almost the speed of just the scope view AND save quite some memory.  But before that happens we should get the basic stuff working with the existing API. Slawek did already all of this experiemental work on that, now we should quickly get the basic stuff done in a reusable fashion.

We can increase the SPI speed a little more (read double it) but in this case we need to convince the XPT2046 to still cooperate with us. This was the main issue why we had to limit the prescaler to /4 instead of the maximum /2. Also some ILI932x did not like the speed but I think most would be able handle it but with the touch controller on the same SPI it seems to be limited may due to some capacitors or SPI wire length issues.

**Comment by sp9bsl** [comment]
Good news Danilo! I haven't try this but i saw in pdf that 9486/88 can do that. It looks that this will be the basic LCD for mcHF/UHSDR now. 
I increased the speed to maximum that spi on APB1 bus can do (divided by 2). The touch controller needs slow down to /64, but it doesn't affect the visual side 'coz all this is very rare. It is the case of touch (clone?) IC, not wires.

Does anyone opened the pdfs correctly?

**Comment by df8oe** [comment]
Hi Slawek,

Linux seems to be unable to open them. They have reasonablew size but results in empry screen. But I can see "3D" in chapters on the left.

Regarding "standard display: Maybe we switch to this types of LCD at OVI40. You can get them without base PCB so that you can use SPI directly or (better) drive them parallel - take a look here: http://www.inanbo.com/index.php/2015-11-03-07-10-45/61-t35frpi So I can design a base PCB for it which fits LCD, carries flat ribbon connector and 2x17pin header which is compatible with OVI40 header. Not very complicate at all. BUT: If this LCD will become standard of mcHF is not in our sphere of influence. That is something Chris must decide and do. But I do not see any problem to switch to new LCD type at OVI40 before start of the delivering of kits/PCBs and so on. And all betatesters who now have the "wrong" LCD can just swap it to the new one.

**Comment by K2AAE** [comment]
@df8oe What is the price for that 3.5 display and what is the controller? I was looking at https://goo.gl/NEiNH3 ILI9488 with the $5 i²C touch screen is $17, I am going to build a shield for my mcHF, use I²C Touch and add flash for fonts, purchased 800mb, fixed and Times Roman, Arial, 35 languages 8 different sizes. All the fonts I saw were English in the git. Moving to flash fonts should free some memory.
STM32H743ZI arrived 2mb flash 1mb sram Ethernet and it works! trying to track down the HAL for it.

**Comment by K2AAE** [comment]
@sp9bsl  "Does anyone opened the pdfs correctly?" I had to 'Turn on 3D' in the preferences of Adobe Reader. No Color just military gray. Looked good ...

**Comment by sp9bsl** [comment]
@df8oe nice one, i think this kind of ribbon cable doesn't need connector. I think it is for soldering directly to PCB. I have some 3" lcds with this kind of ribbon cable, and it looks similar. I designed pcb for cheap raspberry pi lcd that is drop in replacement for 2.8" used in mcHF.  So if there will be anyone who is interested in such board i could send it to pcb factory but the driver code first. Now i changed all the code but i have missed something and my lcd is white - i will commit changes soon after i find what's wrong.
I would like switch to yours OVI40 UI board because of F7 core, but not now.

@Airheadbit I'm glad that pdfs opens, they're gray because of my 3D modeller (ZW3D) i will swith to SolidWorks in a few weeks. Regarding price: raspberry Pi LCD costs about 12$ (ebay, Aliexpress) with free shipping from China [here](http://www.ebay.com/itm/3-5-inch-B-B-LCD-Touch-Screen-Display-Module-320-x-480-for-Raspberry-Pi-V3-0-A-/292128320811?hash=item4404346d2b:g:i0MAAOSwZ4dZJXSd)
Also there is better one with IPS panel for double the price.

**Comment by df8oe** [comment]
@sp9bsl it does not need LCD controlling - but it does not have touchscreen controller. You have to add a board which does correct wiring to 17pin header and must carry xpt2046 and some resistors.

I am actually not able to test new LCD on mcHF ui because I do not have one for testing. But I am ready for testing on OVI40 ui :)

vy 73
Andreas

**Comment by DG9BFC** [comment]
i could test the new lcd on mchf :-) cause mine is not mounted in a case (yet) .... i would just need that adaptor pcb and the lcd (plus a spot where to download modded firmware)

**Comment by s53dz** [comment]
I suppose you will have the FW here eventually.
https://github.com/df8oe/UHSDR/tree/lcd-experimental

73 Bojan

**Comment by sp9bsl** [comment]
I will commit it in a moment, first look of new driver:
![1](https://user-images.githubusercontent.com/17089070/31136588-0f939a54-a869-11e7-98e9-fd1de95c67f3.jpg)
![2](https://user-images.githubusercontent.com/17089070/31136587-0f92e118-a869-11e7-90be-6a743a1c1529.jpg)

Arghh....  Weird view after using my version of firmware for a half year ;-)
73 Slawek

**Comment by db4ple** [comment]
Cool.  
Let us now verify that the new driver did not break the old ones and we can go forward.

**Comment by sp9bsl** [comment]
OK Danilo, also i put some wiring for it:

![wiring](https://user-images.githubusercontent.com/17089070/31137812-9f6c50aa-a86c-11e7-8b0d-f0f15fcee689.jpg)


This is my actual adapter schematic.

**Comment by db4ple** [comment]
Hi Slawek,
now comes the interesting part with git: Please run:

git fetch main
git rebase main/active-devel
And if everything goes smoothly:
git push -f origin

What this does is the following:
git fetch main -> get the current content of the main github repository
git rebase main/active-devel -> apply your changes on top of the active-devel.
git push -f origin -> sent your local branch to your GitHub repository. 

That means your branch will then be in a state where you have essentially all changes in the active-devel and your changes together. This allows us to integrate all changes which happen in active-devel. So the risk of merge problems later are very much reduced. 
In some cases rebase will not work automatically completely and ask you to merge your changes. 
It is not difficult. And if you get in trouble, a simple git rebase abort brings you back to where you started.
Please try this. 
Next time you should do this before you actually send the pull request.
You will see, it is really cool once you get the hang of it.

**Comment by sp9bsl** [comment]
Danilo i have problem:
![image](https://user-images.githubusercontent.com/17089070/31142388-6b4a1f9c-a87a-11e7-9ede-9bd48ee3502e.png)

**Comment by sp9bsl** [comment]
Now i see that with this all the code i wrote disapeared...  Good thing is that i have done backup all the project i have on disk before i played with git...

**Comment by db4ple** [comment]
Hi Slawek,
git will never forget anything. Trust me. 
a) no your code did not disappear. During rebase it is NOT YET THERE. Git will add your changes at the end of the process.
b) What you see is a simple problem. You have to open the file uhsdr_version.h 
and you will find a line  starting with >>> 
a line ====
and a line
<<<<

This is a marker for a merge conflict. You will notice in this case it is the version number.
Here you simply delete part with one of the version numbers and the marker lines (>>>,===,<<<).
Now you have a merged version. 
you now do as git tells you
git add <path_to_the_merged_file>
git rebase --continue

Git rebase may again ask for help with merging. In most case the problems are as simple as the one above. 

Try it. Again, if you are absolutely unsure -> git rebase --abort will bring you back all data as it was before you entered git rebase. No need for a backup (well of course a backup is important but not for this situation).

Git is not the easiest tool but once properly used not very difficult to operate

**Comment by db4ple** [comment]
BTW, you may attach the problematic file here and I can explain you how to proceed.

**Comment by db4ple** [comment]
Hi,
for a little background on what happens during rebase:

https://git-scm.com/book/en/v2/Git-Branching-Rebasing

**Comment by sp9bsl** [comment]
The file:

[uhsdr_version.zip](https://github.com/df8oe/UHSDR/files/1353623/uhsdr_version.zip)

i see point at which it stopped.

**Comment by db4ple** [comment]
Hi Slawek,
first: Thank you for being patient and working this out with us.
Now:

This is the problematic part:
<<<<<<< HEAD
#define 	UHSDR_VER_RELEASE		"116"
=======
#define 	UHSDR_VER_RELEASE		"107"
>>>>>>> releasing new daily build

The top part is what the part from the active-devel branch, the lower part is your content from your branch.
We decide the upper part is more recent, so we take it:

DELETE: <<<<<<< HEAD
#define 	UHSDR_VER_RELEASE		"116"
DELETE: =======
DELETE: #define 	UHSDR_VER_RELEASE		"107"
DELETE: >>>>>>> releasing new daily build

Result is one line (and the rest of the file of course)
....
#define 	UHSDR_VER_RELEASE		"116"
....

Now you do git add with the path to the file, i.e. 
git add mchf-eclipse/src/uhsdr_version.h
and then
git rebase --continue

To be honest with you, I don't know yet what the commits Andreas will do here and if we will run in more problems. I will try the rebase on my own (yes, I can do that, thanks to git and github). Please wait a moment.

**Comment by sp9bsl** [comment]
ok, no problem Danilo, i see the advantages with git use...
Now after i did things you wrote i have problem with eclipse file:
![image](https://user-images.githubusercontent.com/17089070/31146316-9100e3fe-a886-11e7-8e50-595b0ea42775.png)

**Comment by db4ple** [comment]
The solution here is
git rm .settings/org.eclipse.cdt.managedbuilder.core.prefs
git rebase --continue
That gets rid of the unwanted add of the .settings/org.eclipse.cdt.managedbuilder.core.prefs

**Comment by db4ple** [comment]
And you are done.

**Comment by db4ple** [comment]
Almost:
Now get the newly organized branch into your GitHub repo
git push -f origin
The -f is important, since rebase changed the history of your branch and without -f git will not push such a change to the external repo.

**Comment by sp9bsl** [comment]
ok, now i should make pull request to lcd-experimental right?

**Comment by db4ple** [comment]
Yes, please. After Andreas accepts your pull request the lcd-experimental branch in his repository will now contain the latest active-devel code plus your additions regarding the new driver. 

With this rebase repeated (best done before you start working so that you base your changes on the most recent code and then before you push to your external  GitHub repo so that you catch the latest changes again, ensures that in the moment we want to take your changes to the active-devel branch that there will be no unwanted/untested changes during merge.

I guess, you will encounter a few more challenges, feel free to ask for support. Not always this can be given as fast as today, though.

**Comment by sp9bsl** [comment]
Thanks Danilo for help, i learned a lot today. Now i can say that i like it.

73 Slawek

**Comment by sp9bsl** [comment]
Hello All,
Just after i receive acknowledge that everything works OK also on older LCD  i can proceed to the next step (i still have old one, but if it is ok i would not do swap between them to avoid problems). I think this is the moment we should discuss placement of items. I see that there is progress of touch definitions - now are placed on lists not hardcoded, but those lists are not linked with actual screen positions of items. I know that touch panel will be matched at end, but it will also have influence on new definitions.
This are my thoughts ofcourse...

73 Slawek

**Comment by db4ple** [comment]
Hi Slawek,
while others have to confirm that the other modes are still working, yes we have to start thinking. Before that we can do something else: Run waterfall and scope in parallel. You will notice that in the newer code, displaying both at the same time is almost trivial, since we have have unified all but the actual drawing code. Have a look in ui_spectrum.c UiSpectrum_RedrawSpectrum. 

We need to make the waterfall "relocalizable" independent from the scope view. Right now both use the same defines. If split the defines up into scope and waterfall positions, we can change placement easily.

If you could try this, it would get us further. Right now I simply would place the waterfall in the unused area right under the menu buttons. This of course not the final location but will show us that we have done it right.

**Comment by df8oe** [comment]
Hi Slawek,

I am very busy at OVI40 actually so I do only have short time to watch LCD efforts... All must be exactly described so that I can test - no time for code-digging....

If I build a binary using lcd-experimental: will it result in "old LCD code"?

Regarding touchscreen I would appreciate a "translation-table" for every used LCD, where x/y positions of LCD output can directly used for touchscreen, so that you can use same names of definition for outputting to LCD and touchscreen. So you do not have to redefine touch areas for each reginon - only implement the "new translation table". We have not started that yet (we already do have two different LCDs which are supported) but if we add more we should start using more flexible way. I am not sure I have written so that you understand what I am meaning. If not --> ask.

73
Andreas

**Comment by sp9bsl** [comment]
@db4ple OK, i will do that.
@df8oe To compile for old LCD you have to comment out two lines in uhsdr_board.h:

#define USE_GFX_ILI9486
#define USE_DISP_480_320

And regarding translation table: i agree. So this will be the last step i think.

**Comment by df8oe** [comment]
Hi Slawek,

I will build it for standard LCD and report here.

**Comment by df8oe** [comment]
Sadly not functional with old LCD. Only blank screen...

**Comment by df8oe** [comment]
QRX.... maybe my mistake...

**Comment by df8oe** [comment]
Everything ok and working - my mistake...

If I build it for OVI40 and use new driver - can I use new LCD with OVI40, too? I have build an adapter to plug new LCD to OVI40. SPI lines are already defined for OVI40 touchscreen. LCDCS and LCD_RESET I do have wired to correspondent parallel lines.

**Comment by sp9bsl** [comment]
I posted schematic of my adapter here yesterday, few posts above this one. You have to connect the RS pin also. I didn't checked SPI connection on OVI40 if it is the same like in mcHF it should work.

**Comment by db4ple** [comment]
@df8oe : Probably.  We have the spi configured for touch screen use so there is all you need except for the SPI DMA (which is not configured and this is the problem. You need to comment in USE_SPI_DMA (in ui_lcd_hy.c in the start of file)  and hope it works (it should but I have not tested this in a while). Of course this will be slower but should be still okay.

**Comment by db4ple** [comment]
@df8oe and @sp9bsl : Relocation tables: We need for each layout / resolution a set of touch regions and associated actions (this is already partially prepared) see touchactions_normal / touchactions_menu in ui_driver.c (around line 6170).   

For now I would propose to go the simple route: for the new 480x320 layout we simply use a second set of data structures for this setup (#ifdef USE_DISP_480_320) and think about the translation tables later.

Later we need for each combination of touch driver/resolution or even worse every display/touch driver/resolution combination a translation into a unified model for that resolution so that for a single resolution we can get away with a single logical coordinate system for touch regions.

**Comment by sp9bsl** [comment]
@db4ple We should go this way, it's ok now.
@df8oe I see that APB1 for SPI2 runs faster by 12MHz (54 MHz vs 42 in f4) but i think it should work. If not i will modify prescaler for OVI40.

**Comment by df8oe** [comment]
Sounds very good. I will test it probably this evening.

**Comment by sp9bsl** [comment]
@db4ple  I added waterfall display in free space as you described. I see big differences in code (in plus) so it took less than a 5 minutes :) I will now read about and try "hardware" implementation of scrolling in this lcd. Waterfall now moves upward (as original) but it looks strange, i think we should add a switch in menu that will allow to change scroll direction up or down.

73 Slawek

**Comment by db4ple** [comment]
That was what I expected. 

HW scrolling would be very nice since we don't have to shuffle huge amounts of data through the SPI. Looking forward to it. Hope the hardware will work as described in the datasheet. Maybe one of the libraries such as the one mentioned by @Airheadbit will provide examples how to do it right.

**Comment by DG9BFC** [comment]
hi slavek
with your circuit diagramm it should be easy to hand wire an adaptor for the rasi display :-)
making a pcb adaptor is another story ... but i have a defective 2.8 display where i can use the pcb from ...
glue a connector for raspi display to it ... some double adhesive tape ...and the display
hmmm just thinking how i can "hack" such a board together (even if it may look a bit ugly style)

**Comment by df8oe** [comment]
Hmmm. I have built lcd-experimental for F7. USE_SPI_DMA was already defined. OVI40 is booting but result is only a blank screen. green LED turns on, red follos, then both stay some seconds lit, then red gets dark. But screen remains blank...

**Comment by sp9bsl** [comment]
@DG9BFC You can ofcourse use the pcb from old lcd, but you first must remove touchpanel ic, and the rest compnents...
@df8oe You should check with oscilloscope signals at CS and RS (i use D11 as CS pin). When you press BND+ do you hear relays clicking? If yes the program works but there is something wrong with lcd connection or any kind of incompatibility with F7. You can change prescaler to /4 in UiLcdHy28_SpiInit inside #ifdef USE_GFX_ILI9486 in the beginning of the function body

**Comment by df8oe** [comment]
There is audio out, and af gain is working. FW is running in "stealth mode" :) Before measuring and digging into trouble shooting (no time for that right now):
Does OVI40 use same GPIOs for parallel LCD as mcHF? All signals are present on 34p header and I 100% sure have wired correctly. I tried next build - same result. Maybe it is a "simple trap" like a different GPIO usage.

**Comment by s53dz** [comment]
Off topic (HW):
could you, please, confirm the type of 3.5" LCD you are testing in this thread:
http://www.ebay.de/itm/Waveshare-3-5-inch-320x480-Touchscreen-TFT-LCD-D

TNX 73 Bojan

**Comment by sp9bsl** [comment]
Hi Bojan,
i use [this one](http://www.ebay.com/itm/3-5-inch-B-B-LCD-Touch-Screen-Display-Module-320-x-480-for-Raspberry-Pi-V3-0-A-/292128320811?hash=item4404346d2b:g:i0MAAOSwZ4dZJXSd). But there is Waveshare original and few clones, all of them points to the same firmware for raspberry so i assume they are all compatible with my one.

73 Slawek

**Comment by sp9bsl** [comment]
Andreas, i see that you have different pins for parallel iface on OVI40, you should connect CS to D9 of lcd and RS to PD11 of STMF7, unfortunatelly you do not have it connected to LCD socket (your RS pin compared to mcHF board), or simply change definitions in uhsdr_board_config.h to right ones for your board - i see there is pin mismatch within #ifdef for F7.

73 Slawek

**Comment by db4ple** [comment]
Hi Andreas,
read carefully what I wrote "comment USE_SPI_DMA" (or not define/undefined), i.e.
'// #define  USE_SPI_DMA'
so that we run without (!!!) DMA. SPI DMA has not been configured for the OVI40 since we had only a parallel display...

A better solution (not good, since we should get DMA also on the OVI40) for now would be

#ifndef USE_GFX_ILI9486
#define USE_SPI_DMA
#endif 

Please be aware that you have to build the bootloader with the same options to see something during boot. For testing, no need for bootloader output.

**Comment by sp9bsl** [comment]
I have bad news. There is no possibility to do scrolling  "in hardware". I spent two days on it wondering what is wrong, then read some and i found that all lcd controllers which have "hardware scrolling" uses it in vertical direction, but word "vertical" means vertical in the recommended by design direction of view. We use lcd rotated by 90 degree, so we can do only hardware horizontal scroll...  So we must stay at large memory usage. It's a pitty. There is other way but will not add a progress to this project i mean read from lcd and rewrite...

73 Slawek

**Comment by db4ple** [comment]
Hi Slawek,
this is indeed very frustrating. Yes, reading and writing would be possible but that would take time.
So I think we have to go with the memory use. But since we are going to do it the right way, users with smaller displays will not have to pay the price for it and all others will have to have or get a cpu with larger ram mounted before switching to the larger displays.

Here we can probably see that these displays have its origins in the mobile phone industry (portrait vs. landscape).

**Comment by df8oe** [comment]
Hi Slawek,

I think at this stage there is reasonable edge of mcHF reached. You have to decide what to implement and what not freedv/further digital modes vs. LCD with spectrum&waterfall simultaneously... But we can do further investigations because on OVI40 there *is* enough RAM for doing all this. There is much to do. Not only the driver, but a new layout of screen, where buttons and text cannot be much smaller than at existing 3.2" LCD because of it would be much more difficult to read. Bur all this future will show.

**Comment by db4ple** [comment]
Well, I think there is no problem to keep the mcHF@192k RAM working (my mcHF is one of those, BTW), it will just not be able to use the larger display.

**Comment by df8oe** [comment]
That's what I meant...

**Comment by sp9bsl** [comment]
I think we should disable freedv (just possibility to turn it on) in use with larger display in devices with 192k RAM or warn user to not to do that. When larger memory is detected user could be able to choose it.

**Comment by db4ple** [comment]
I will look into the enabling the "FreeDV disabling". Right it does not work since it shares code with the new NR (the buffers).

**Comment by df8oe** [comment]
We are reaching another edge of mcHF, too. In the beginning of the project mcHF was delivered with 512KB MCU. As I started building my mcHF and stated it would be wise to use 1MB instead it lasts ~half a year until this was done at all new mcHFs. So there is an unknown amount of mcHFs which only do have 512KB MCU. We already scrape at this limit and with implementation of new LCD driver we definitly would leave 512KB. So I think I will start a survey next days at Yahoo how many users do have 512KB MCU... Detection of MCUs with 256K Ram is already implemented. But there will only be a handfull mcHFs which uses already this MCU.

**Comment by sp9bsl** [comment]
Andreas, Yes you're right. But will all they need this improvement? They need to choose to do this modification and don't use freedv or use it with old lcd. Same like someone have to decide to upgrade to new firmware (which will excess 512kB of flash) or not. There is no way to accommodate everyone i think. Yes, it would be good to start the suvey and know. Sooner or later you will reach 512kb limit anyway.

**Comment by df8oe** [comment]
So it is. But interesting to know who is affected by this...

**Comment by db4ple** [comment]
Hi,
I just built a OVI40 which is without FreeDV around ‭316864‬ bytes Flash. There is significant room if FreeDV is left out for quite some functionality and also RAM usage,
For the mcHF the flash usage without FreeDV is around 260kByte (!)

**Comment by sp9bsl** [comment]
What is the difference between mcHF and OVI40, HAL for F7?

**Comment by db4ple** [comment]
OVI40 -> F7 -> HAL F7
mcHF -> F4 -> HAL F4
My numbers are Eclipse builds using the same compiler and a static DSP Lib.

The FreeDV on the F4 uses a few F4 specific tricks which keep the performance high (it basically uses in some places the DSP Lib) , this increases flash use dramatically in Eclipe builds but keeps RAM use low. In makefile builds this is less pronounce, here we use the gcc Link-Time Optimization across all code including DSP Lib (not possible with Eclipse on Windows). This makes the mcHF F4 makefile builds smaller even when we have FreeDV in.
On the F7 we run the normal FreeDV code since we have enough RAM to do so. The FFT Lib of FreeDV is quite well performing but needs RAM tables.

**Comment by sp9bsl** [comment]
Thanks Danilo. If we decide to use new LCD with STM32F405VE we should have conditional build for disable FreeDV anyway. So i propose:

1. F405 - with FreeDV and old LCD (normal mcHF build), (we already have almost no room in FLASH.)
2. F405  -without FreeDV and new LCD, (some free space in flash after FreeDV remove)
3. F7xx, and H7xx - with all goodies (new LCD+FreeDV). 

The last option is for new OVI40 board of course.

**Comment by s53dz** [comment]
Hi,

I suppose we have to be more precise here. 
My case was like this:
mcHF with STM32F407**VE**T6 (**512kB, 196kB**), which I changed to STM32F429**VG**T6 (**1MB, 256k**).
So, the suffix is the defining one (beside the type F4 / F7, of course).
VE = 512kB, VG = 1MB, VI = 2 MB.

73 Bojan

**Comment by sp9bsl** [comment]
Hi Bojan,
i especially write VE to indicate 512kB of flash, i know naming convention:) By the way, i see that chinese clones also use 512kB flash core, there will be a lot of people own them.

73, Slawek

**Comment by s53dz** [comment]
Perhaps it is a moment when we should explicitly list different mcHF (UHSDR) modes of interest for the present with a look to the future. So, beside the SSB and CW, **all other modes should be switchable either in a makefile and/or in a menu**, shouldn't they?

**My present** sight is quite simple:
SSB
CW
(AM) - possibly just for listening to a BC
(FM) - not on HF
WSJT(X) (JT65, JT9, FT8, ...)
PSK31
RTTY

73 Bojan

PS: I suppose this is already off-topic, but originating from a bigger LCD option.

**Edited by Bojan**

**Comment by DD4WH** [comment]
Hi!
add to the list:
SAM - synchronous AM demodulation
IQ - Stereo reception by listening to I on the left and Q on the right ear
Stereo SSB - LSB left ear, USB right ear
Stereo WFM - wideband stereo FM [only if the codec can be made to work out-of-specs with 192ksps]
and yes, NFM (narrow FM) is being used on HF (10m)
73 Frank

**Comment by DG9BFC** [comment]
hello frank hello bojan
do not forget freedv
mchf is the first trx with built-in freedv and we should not "throw that away"
tony made a short video showing freedv in action on an mchf
 https://www.youtube.com/watch?v=MGIjoGcT2Ek
73 sigi

**Comment by sp9bsl** [comment]
Hi Sigi!
We don't want to throw away FreeDV. There is no option to have both big LCD and freeDV in F4 core with low RAM (people mostly have this option). In F7/H7 core there will be both.

73 Slawek

**Comment by DG9BFC** [comment]
ahhh i see ... freedv only excluded if user has small ram and big lcd .... ok on that ...
so if user wants to keep freedv and have big lcd he has to swap mcu to have more ram (if he has a small ram mcu) ... ok ... i can live with that :)

**Comment by db4ple** [comment]
Hi,
I am not sure that we should have this discussion regarding how to fit all functionality here. If you want to discuss this, please open another issue dedicated to this topic, since it is not only the larger display but also now modes etc. which can fill the flash and the discussion what to include when has not much to do with the actual work on the larger LCD. This is an important discussion but to a certain extent also pointless if we work to enable relatively easy compilation of the different functionalities into working firmware variants. @Airheadbit seemed to be interested into setting up a online "compilation services" which would then produce the right binaries. But even without that, it is almost trivial to build different configurations (if we make the relevant parts "removable" during compilation).
It is not trivial if the number of switchable features (and their dependencies) get out of control but that is not yet the case.

**Comment by db4ple** [comment]
@sp9bsl: Please commit your current state as usual. We haven't found a good solution for the reappearing conflicts, but as long as it is only the version number, we will have too much trouble to fix that...

**Comment by sp9bsl** [comment]
Hi Danilo, i will do it this evening. At morning i tried to optimize code, first started with waterfal drawing, but i see that my style has no profit over usual code, so i reverted changes. So this is real bottleneck, now when drawing 60 lines it tooks about 23ms (SPI with DMA) still for 256 points. 
You calculated the ram usage for every line of new waterfall data, but i think i can use the trick you can see on my video. I forgot to say that every line of waterfall is doubled. This is relatively high resoultion lcd so it is not noticeable by "normal" eye that my function draws every line twice. It saves memory by twice but not the drawing time (again SPI transfer over DMA). Summary: we have (height/2)*480 bytes of data for waterfall. I think i will add this to the next commit, so what do you think about it?

73 Slawek

**Comment by sp9bsl** [comment]
Now i think that would not be good to have dynamic memory allocation for waterfall? And take what is available for current STM? Just simply use malloc()?? And use it with line reperating x2, x3, x4..... Just thinking.

73 Slawek

**Comment by db4ple** [comment]
Hi Slawek,

Doubling to keep Ram usage low is a good idea. On higher RAM machines we can use full resolution. If we have the half number of lines and twice the width we have same ram usage as before.

We have to see how fast we can push the SPI bus.

73
Danilo

**Comment by sp9bsl** [comment]
I pushed it to the limit in my last commit. APB1 /2... We could change SPI number but it is hard to do now. Only SPI1 have high speed of 84/2 MHz. Its pins are connected to M2 pushbutton and E1 encoder.
In parallel interface there will be no bottleneck with DMA usage (mem to mem transfer over FSMC).
I think i will try assembler routine for "manual" transfer over SPI pins. I see in HC4094 pdf (i have it on back of my LCD in serial to parallel interface) can go up to 80 MHz.

**Comment by db4ple** [comment]
Hi,

malloc is available. Currently used by FreeDV only, but can be used for anything. However , I am not sure why you intend to do. I would simply allocate the data structure for height/2 * 480 or height * 256. In case of Double line, we simply send two times the same data.
If this was what you meant, malloc is good choice.

73
Danilo

**Comment by db4ple** [comment]
Hi,
wanted to let you know that we have another graphics driver coming to the UHSDR platform: RA8875. Thanks to @gmtii.  See here: https://github.com/gmtii/UHSDR 
I volunteer to do the integration of the low level graphics driver into a unified piece of code. Low Level means the ui_lcd_hy28.c file. 
Result would be a source file which is able to drive all three kinds of display running the 320x240 code.

Now the important part: In order to avoid wasted effort I think it would be best to align the work on the higher level changes (UI layout and spectrum width). Although there are differences (480x320, no hardware scrolling vs. 800x480, hardware scrolling and more) there are also a lot of overlapping changes (we need to support multiple fft sizes, multiple ui element layouts). Coordination and collaboration is the key. Two sides producing changes to the same area of code possibly using different approaches will not work.
We can either have two iterations of integration (one display after the other) or you share the work.

**Comment by sp9bsl** [comment]
Hi Danilo, i'm ready for team work (we alredy have this). I had almost no time for coding in the last weekend (family). I tried to optimize transfer on SPI2, with no positive result. 
I have few questions:
Do you have any description/video how looks working RA code? 
Will you add the changes to support the RA driver? 
Do you want to throw away the concept of 480x320 display?

**Comment by sp9bsl** [comment]
Ok, i was too fast with creating questions :) At @gmtii profile there are two links to videos (old). I saw this few monts ago but i didn't pair them with gmtii user. I see that he commented my video last weekend.
Anyway, can you describe the collaboration at this level, and which one have higher priority?

**Comment by db4ple** [comment]
Hi Slawek,
a) https://github.com/gmtii/UHSDR/blob/active-devel/mchf-eclipse/drivers/ui/ui_driver.c
The driver is not my work, to be clear about this. All credit goes to Esteban.

b) I would do so so that all 3 drivers are working (one at a time). I already did a test merge and it looked quite good.

c) Absolutely no. The 3.5 480x320 is a good display compromise, much higher resolution, slightly larger screen size. The RA8875 display is a 5 inch 800x480 display (SPI and parallel) with a much more advanced display controller and also offered with nice capacitive touch controllers but it will not fit into the concept of a small, low power device. When active, this thing draws naturally much more power, the case needs to be quite a lot bigger etc. For a shack it is quite nice, also touch due to larger size is more convenient I think, and there are even larger displays (7") with this controller but all of this makes it not ideal for QRP outdoors I would say. So I tend to see it as an interesting challenge to support all 3 displays which should cover more or less the range we can support with this hardware concept. 
Going larger would be much easier with a dedicated controller board handling most of the UI stuff independently. But then we are talking a completely different hardware and software architecture.

Regarding the collaboration, Esteban ported some of this work to the UHSDR (thanks for that, BTW!) but as you he also found out, it is quite easy to modify a UHSDR system to one's liking but keep the existing solution and adding something fundamental as a different UI layout is a challenge in the UHSDR firmware (and not only there). So we have a bunch of possible futures as a set of individual modifications and someone need to put all of this together (which is not a simple merge/change).

The whole benefit is that once done properly many new features will be available to all of us. Which is a cool thing, isn't it?

**Comment by sp9bsl** [comment]
Yes, of course it is! I saw this lcd in Tulip trx and it looks and works great (8" capacitive version) but its price... Anyway everyone can choose. This will be great chalenge to have it all supported in one source,  i think it is possible to do. I will commit soon my concept when i have it working (memory mapping). I was out of computer this weekend but not my mind :) There was a lot of thinking about pointers, arrays etc....

**Comment by gmtii** [comment]
Hello,

It is a pleasure to participate in a project like UHSDR / MCHF. Many thanks to all who keep the project so alive.

About the above comments: RA8875 is a driver that can handle lcd screens of various sizes and adds powerful functions that can speed up tasks that are slow in software such as waterfall and acceleration of 2D objects.

It is true that it was used in the first versions of TRX, but, if I am not mistaken, now uses screens with LTDC parallel interfaces. It is also true that for small resolutions like 320 or 480, these interfaces are not necessary, and are more thought for bigger screens.

Before implementing code for RA8875 in MCHF, I also teste  ILI9488 and SSD1963 drivers.  I have also experimented with LTDC interface on the stm32f429 discovery board, but it uses too many cpu pins for practical use.

Undoubtedly the commitment of appropriate size and consumption are screens of 480 x 320 pixels.

In this sense the ILI9325 code is almost 99% compatible with ILI9488 like the ILI9486, so it could also be a potential and easy candidate to incorporate into the MCHF code.

So I can try to make a version of UHSDR that supports these screens that can be merger later. Of course I´m willing to coordinate this efforts with Slawek.

73.

Esteban.

**Comment by sp9bsl** [comment]
Hi Esteban,
i'm glad to see you here. I implemented driver for ILI9486/88 in lcd-experimental branch. It is the easiest part of work. Now comes all the compatibility issue. We have here STMs with 192, 256 and 512 kb of RAM. Also different resolutions: 320x240 vs 480x320, and finally your's new 800x480. The pros of your display begins with vertical hardware scrolling capability in landscape position (ILI9486/88 do not have this as i wrote few posts above). So it should be easier than with 480x320. The big advantage here is no memory requirement for waterfall caused by hardware vertical scrolling. So i think we all have to discuss screen positions and size of items. I think the key for 3.5 display is to first see the items on this lcd using existing code from lcd-experimental and then discuss its useability (all are a bit smaller). I think same or worse  problem will be on yours 5" lcd.

@db4ple I think i resolved waterfall ram problem for f4/f7. It should be tested now because my f4 have only 192kB of RAM. If there is noone with 256k of RAM i could prepare test for my core although. This can be checked in my next commit.

Now we have options:
1. core F4, display 320x240 code without changes (normal mcHF)
2. core F4, display 480x320, autodetection of ram size and for 192kB use of array declared in CCM ram and double each line of waterfall, for 256kB and higher allocate new array for waterfall in normal ram (not CCM) and use single lines
3. core F7, display 480x320, autodetection of RAM (we always have >256kB of ram in f7) allocate new array for waterfall in normal ram (not CCM/DTCM) and use single lines.

All memory maping is done by pointer in sd structure which points to waterfall data memory.

So, please upload lcd-experimental to board with new 3.5" lcd and check if it works and then decide what we do with sizes of items.

73 Slawek

**Comment by sp9bsl** [comment]
@gmtii, i have question regarding spectrum display. I think it will be faster to ask You than dig in your code. Did you use ane kind of interpolation/summing to squash up the 1024 fft to 800 pixel wide screen?
Or just show part of the array. In my case i did the second because one bin of 512 fft contains information of about 93Hz so when 32 pixels are not shown i lost 3kHz of visible spectrum. In your case it would be around 10kHz (!). I think i will use some kind of floating indexing and adding to squash it and have full 48kHz, but if you did it we could use it in the same way.

73 Slawek

**Comment by gmtii** [comment]
@sp9bsl I did a function that does exactly an interpolation and that could be adapted to the width of LCD screen that we use, as well the fft size. I guess 512 fft size is ok for the 480px LCD.

Here is the code:

[https://github.com/gmtii/UHSDR/blob/e94850bd80660bf0bc1e435a443cfbe06c621bab/mchf-eclipse/drivers/ui/lcd/ui_spectrum.c#L836](url)

On the other hand, and following the conversation of the screens, I think that the SSD1963 controller has hardware scroll function and there are even cheaper 3.2-4.3 screens than the standard hy28. The difference in memory usage and speed by having scroll hardware is great.

Esteban.

**Comment by sp9bsl** [comment]
Can you provide a link to lcd with SSD1963 with that price?

**Comment by gmtii** [comment]
[https://goo.gl/qAYFUi](url)

**Comment by sp9bsl** [comment]
About conversation on lcds:
I calculated ram space for use with simple RGB interface and LTDC, and there is enough ram in F7/H7 to keep all display ram inside without use of SDRAM up to 800x480 in 8 bit LUT mode. In this option we do not have any transfers on SPI/parallel just simple ram. External RGB interface can be still 16 bit if we connect it properly. The drawback is the noise (there is need to have short traces, and use some EMI reduction techniques), and different pinout.

**Comment by sp9bsl** [comment]
@gmtii thanks, i found it on ebay also for the same price. It is twice of mine and have 420x272, but larger size 4.3". People (maintainer) reading this have to decide where we go.

**Comment by yo2ldk** [comment]
If you can do it, larger one is better, especially for digital modes decoders option
7" is too big, 2.8 too small; but 4.3 to  5" is optimal from my point of view.

**Comment by df8oe** [comment]
Hi Slawek,

I think rebasing across different branches does not work. I have unified my active devel with my lcd-experimental and my lcd-experimental contains all your changings. up to your last pull request. S I think we can solve the merging conflicts if you do rebase your lcd-experimental to my lcd-experimental (not to my active-devel). Just an idea...

**Comment by sp9bsl** [comment]
Hi Andreas,
i will do it afternoon when i'll be back at home.

**Comment by sp9bsl** [comment]
Esteban, 
i think that when i finish my concept i step aside to let you introduce your changes. Now i spent some time working on it and want to finish it. I would be happy then to modify my effort to meet your requirements also with new lcds. Is it ok for you?

**Comment by gmtii** [comment]
Ok, no problem.

**Comment by db4ple** [comment]
@sp9bsl , @df8oe :
Let us clarify: We now make sure the merge of Andreas did the right thing. We need to do this. 

And then we run the whole operation in the right way:

@sp9bsl: 

 1. You create a branch in your local repository which takes the current df8oe:active-devel as starting point.
  1. You push this branch into your GitHub repository
 
  1. You do your work, you run as often as possible rebase against df8oe/active-devel. 

Now you are in charge of this work. Everyone who wants to contribute to the LCD work must use your branch as base for their work and use this as rebase source instead of df8oe/active-devel  and work against your branch. That means, you will have to accept pull requests and merge them or reject them.

@df8oe:
1. In order to promote this work, you create a branch in the public GitHub which follows the progress of Slaweks branch exactly, but you will not accept any pull request for this branch but sent all users to Slawek for merge.
1. At some point you accept the pull request from Slawek into df8oe/active-devel, which will be easy peasy since he has everything based on df8oe/active-devel.

That will work without problems if everyone sticks to the simple principles (regular rebase, where to accept pull requests) outlined above, this is basically exactly as we have been doing it for more than 18 months now.

@sp9bsl, @df8oe: Please agree or disagree explicitly. It is just a proposal from my side, but since I have been working this way with git several times, I know it will work.
 

73
Danilo

**Comment by sp9bsl** [comment]
It is little complicated to me because of almost lack of experience with github, so as before i will need some help from you or Andreas, but in generally yes, i agree. I will do a simple change to code to test it today.

**Comment by df8oe** [comment]
Hi Danilo,

I strongly want to have a working solution for all sites and I am willing to do everything that is neccessary. So of course I agree to your proposal. Maybe it is interesting for others, too I want to discuss my steps here in public.

What do you mean with "...create a branch in the public Github which follows the progress..."? Shall I discard the existing lcd-experimental at Github and in my local repository? And if so: If I create a new lcd-experimental it has exactly the same state as active-devel. How can I follow the progress of Slaweks branch if I get no pull request from him? Must I step through my local lcd-experimental, merge Slaweks progess there and push it to GitHub?

@sp9bsl 
We will keep it as simple as possible for you. I have merged all your progress into active devel so that there is nothing special to change for you. My last post where I recommended a rebase against lcd-experimental is now wrong. Do everything as you did before: rebase as often as you can against active-devel, do your work and push it to your Github.

**Comment by db4ple** [comment]
@df8oe: For the sake of keeping the history, I would propose that Slawek uses a new name for his branch.
Alternatively, to keep the history you simply create a branch lcd-experimental-obsolete based on the head of lcd-experiemental and then you can safely delete the lcd-experimental and "reuse" the name. 
Same applies to Slawek. But honestly, I think we use a new name, and that's it.

**Comment by df8oe** [comment]
I would use the name "wip-4-new-lcds"...

**Comment by db4ple** [comment]
@sp9bsl: It is important that you do one thing right: Create a new branch based on df8oe/active-devel. Do not continue to use your existing branch lcd-experimental. If you continue to use the current branchs conentet, the mess will continue like we have it now. Since Andreas merged all your changes into active-devel, it should be a no brainer. If you have further changes, well, you have to merge (!) these changes into your new branch. Not a big deal.


73
Danilo

**Comment by db4ple** [comment]
Name is fine, we could use that as a general idea (wip-.... ) for these branches. But saying this, maybe we should include the  owner "wip-sp9bsl-4-new-lcds" so it is a bit more clear where the pull request should go to and who takes the "ownership".

**Comment by df8oe** [comment]
Very good proposal. I would start this way (but do it exactly when all have agreed)...

**Comment by sp9bsl** [comment]
I agree too.

**Comment by sp9bsl** [comment]
something is not ok, i don't see my changes, 
I see...  I created branch in my repo, but should i in the main project?

**Comment by db4ple** [comment]
Where can't you see your changes? Which changes?

**Comment by sp9bsl** [comment]
I created branch from actve-devel on my repo. Is that ok?

**Comment by db4ple** [comment]
No.
Instructions:
Assuming you have no local changes anymore in your current branch (`git status` will tell you).
I assume you named the DF8OE GitHub repository reference  'main' in your local repository.
`git remote -v` tells you in the first column the name of the remote repository followed by its url.

1. switch to active-devel in your local repository: `git checkout active-devel`
1. get the most current information from the remote GitHub of df8oe called 'main' into your local repository: `git fetch main` . Background: this does not change any of your own local branches but updates or creates the internal copy of df8oe's GitHub repository (these copies are the "remote" repositories). 
1. update your local branch active-devel to the state of main/active-devel: `git rebase main/active-devel` Now your repository is identical to the one outside if you made not changes to active-devel locally.
1. create the new branch wip...  based on your current status in the currently checked out branch active-devel : `git checkout -b wip...`
1. push this branch to your remote GitHub (probably named 'origin', if not, see above): `git push origin` . This will fail since the branch is not yet in your remote repository, but git is so kind to give you the proper  command line in its error response. Use that. 

Done. Rest for a while, and admire the result...

**Comment by sp9bsl** [comment]
I think it works. As usual:  thank you Danilo! No problems this time.

73 Slawek

**Comment by db4ple** [comment]
Looks good to me. BTW, github tells you, when your GitHub branch is behind DF8OE's active-devel, here you can see, that it is a good time to rebase the local repository and to push the result to your github.  
In the end you will see something like "10 commits ahead of df8oe:active-devel". That means you did alright, since all your changes have been "virtually" made on top of the most current active-devel state which in turn means Andreas has no trouble to merge without conflict. You will see.

**Comment by sp9bsl** [comment]
Hi all,
as all work with new lcds have been moved to new branch, please add questions/reports to issue tab in my repo. [Here](https://github.com/sp9bsl/UHSDR/issues).

73 Slawek

**Comment by K2AAE** [comment]
In the latest: 
#ifdef USE_DRIVER_RA8875 
  #define USE_DISP_800_480 
#endif 
You know there are displays that use the RA8875 at 5" 320X480

**Comment by sp9bsl** [comment]
@K2AAE Can you provide a link to this lcd (ebay, aliexpress or other)?

**Comment by K2AAE** [comment]
https://goo.gl/23qe33
They have many displays and controllers - click top left TFT and pick you're size.

From the RA8875 document
 Supporting TFT Panel Size:
 800x480 Pixels 2 Layers @ 256 Colors.
 800x480 Pixels 1 Layer @ 64K Colors.
 480x272 Pixels 2 Layers @ 64K Colors.

**Comment by db4ple** [comment]
Of course most controllers can support different resolution.
But at the moment we have a clear connection between both. 

https://github.com/sp9bsl/UHSDR/blob/fce6a9e40ffc5f2c893e4bd31a46a53e81124de9/mchf-eclipse/hardware/uhsdr_board.h#L40  reflects that.

Be logical separating the driver and the resolution into separate defines now we will be later able to "freely" combine resolution and driver (if we use the ifdefs in the right places in the code).

**Comment by WM0I** [comment]
When you post new firmware updates will the older LCD have it's own firmware update? I hate the idea of not getting new firmware.

**Comment by df8oe** [comment]
We will release different versions of firmware if other LCDs are implemented fully. Actually only 320x240 LCDs like HY28B or HY32D (and it's "SPI daughter") are supported by released binaries. Other LCDs are in heavy development - there are NO binaries released yet and I think it will last some weeks if new LCDs are fully integrated. If it is so, we will make an announcement, there will be instructions for connecting them and there will be different binaries. Of course old LCDs will be supported in future 100%. But some new goodies which will only work at new LCDs of course will not come into 320x240 binaries...

**Comment by sp9bsl** [comment]
I think we can close this long issue. Driver works and all new work is focused on unification of firmware. @df8oe: what do you think?

**Comment by db4ple** [comment]
👍
